ProjFS.Mac/PrjFSKextTests/MockVnodeAndMount.hpp

```diff
@@ -63,10 +65,13 @@ struct vnode
 
     void SetGetPathError(errno_t error);
     void StartRecycling();
+    void SetAttr(int attr);
```

Comment:
Thoughts on my strategy here

SetAttr I believe is valid as a setter since you flip the bit on a file.  I.e. this is change that can happen once the vnode is created.

SetGetAttrReturnCode gets the job done, but it will become verbose if we need similar on multiple functions.

Reply:
> Thoughts on my strategy here
> 
> SetAttr I believe is valid as a setter since you flip the bit on a file. I.e. this is change that can happen once the vnode is created.

I think it could be named better, as `va_flags` is not the only attribute that `vnode_getattr` can return. So something along the lines of `SetAttrFlags` might be clearer. However, your second point is excellent and actually leads me to another suggestion for the attributes too:

> SetGetAttrReturnCode gets the job done, but it will become verbose if we need similar on multiple functions.

Yeah, this really won't scale well. I'm wondering, perhaps we should implement the mock return codes as a struct?

```
struct VnodeMockErrors
{
    errno_t getpath;
    errno_t getattr;
    // …
};
```

And then just add a public member variable of that struct type to our mock vnode. They'll all default to 0, and you can then easily set the accessor function errors without tons of boilerplate:

```
testVnode->errors.getattr = EINVAL;
DoStuffToVnode(testVnode.get());
```

Similarly, I'm thinking we could just have a public member of type `vnode_attr` where we can set whatever attribute values we want without writing setters for each of them.

Reply:
@pmj I like that suggestion a lot!
I'll do that in the next revision.

Reply:
> I'll do that in the next revision.

Note that my existing `inode` member variable, getter, etc. also would fall under this (it's the `va_fileid` attribute) so you can get rid of all of that too!

Reply:
I pushed this up and I like the look.  Let me know what you think.

I left SetPath as is since it does more work than just set a value.  I did update the error return code for getpath.